File: casper/src/test/scala/coop/rchain/casper/batch1/MultiParentCasperFinalizationSpec.scala
package coop.rchain.casper.batch1

import cats.syntax.all._
import coop.rchain.casper.helper.TestNode
import coop.rchain.casper.helper.TestNode._
import coop.rchain.casper.protocol.BlockMessage
import coop.rchain.casper.util.ConstructDeploy
import coop.rchain.p2p.EffectsTestInstances.LogicalTime
import coop.rchain.shared.scalatestcontrib._
import monix.execution.Scheduler.Implicits.global
import org.scalatest.{FlatSpec, Inspectors, Matchers}

class MultiParentCasperFinalizationSpec extends FlatSpec with Matchers with Inspectors {

  import coop.rchain.casper.util.GenesisBuilder._

  implicit val timeEff = new LogicalTime[Effect]

  val genesis = buildGenesis(
    buildGenesisParameters(bondsFunction = _.map(pk => pk -> 10L).toMap)
  )

  "MultiParentCasper" should "advance finalization monotonically in round robin" in effectTest {
    TestNode.networkEff(genesis, networkSize = 3).use { nodes =>
      for {
        deployDatas <- (0 to 7).toList.traverse(
                        i =>
                          ConstructDeploy
                            .basicDeployData[Effect](i, shardId = genesis.genesisBlock.shardId)
                      )

        // Round-robin block production across 3 validators with propagation.
        // With multi-parent merging, each block may have multiple parents
        // (one from each validator's latest message).
        block1 <- nodes(0).propagateBlock(deployDatas(0))(nodes: _*)
        block2 <- nodes(1).propagateBlock(deployDatas(1))(nodes: _*)
        block3 <- nodes(2).propagateBlock(deployDatas(2))(nodes: _*)
        block4 <- nodes(0).propagateBlock(deployDatas(3))(nodes: _*)
        block5 <- nodes(1).propagateBlock(deployDatas(4))(nodes: _*)

        // After 5 blocks in round-robin with 3 validators (equal bonds),
        // finalization should have advanced past genesis.
        lfbAfter5 <- nodes(0).casperEff.lastFinalizedBlock

        block6 <- nodes(2).propagateBlock(deployDatas(5))(nodes: _*)

        lfbAfter6 <- nodes(0).casperEff.lastFinalizedBlock

        block7 <- nodes(0).propagateBlock(deployDatas(6))(nodes: _*)

        lfbAfter7 <- nodes(0).casperEff.lastFinalizedBlock

        block8 <- nodes(1).propagateBlock(deployDatas(7))(nodes: _*)

        lfbAfter8 <- nodes(0).casperEff.lastFinalizedBlock

        // Verify finalization has advanced past genesis
        _ = lfbAfter5 should not be genesis.genesisBlock

        // Verify finalization advances monotonically (block number never decreases)
        _ = lfbAfter6.body.state.blockNumber should be >= lfbAfter5.body.state.blockNumber
        _ = lfbAfter7.body.state.blockNumber should be >= lfbAfter6.body.state.blockNumber
        _ = lfbAfter8.body.state.blockNumber should be >= lfbAfter7.body.state.blockNumber

        // Verify finalization has advanced meaningfully by the end
        // With 8 blocks from 3 validators, LFB should be well past block 1
        _ = lfbAfter8.body.state.blockNumber should be >= 2L

        // Verify all validators agree on finalization
        lfbNode1 <- nodes(1).casperEff.lastFinalizedBlock
        lfbNode2 <- nodes(2).casperEff.lastFinalizedBlock
        _        = lfbNode1 shouldBe lfbAfter8
        _        = lfbNode2 shouldBe lfbAfter8
      } yield ()
    }
  }
}

File: casper/tests/batch1/multi_parent_casper_finalization_spec.rs
// See casper/src/test/scala/coop/rchain/casper/batch1/MultiParentCasperFinalizationSpec.scala

use crate::helper::test_node::TestNode;
use crate::util::genesis_builder::GenesisBuilder;
use casper::rust::casper::MultiParentCasper;
use casper::rust::util::construct_deploy;
use crypto::rust::public_key::PublicKey;
use models::rust::casper::protocol::casper_message::BlockMessage;
use std::collections::HashMap;

// TODO: Round-robin finalization concept no longer applies with multi-parent merging.
// Scala deleted this test in PR #288.
#[tokio::test]
#[ignore = "Round-robin finalization concept no longer applies with multi-parent blocks"]
async fn multi_parent_casper_should_increment_last_finalized_block_as_appropriate_in_round_robin() {
    fn assert_finalized_block(node: &TestNode, expected: &BlockMessage) {
        let last_finalized_block_hash = node
            .block_dag_storage
            .get_representation()
            .last_finalized_block();

        // Scala uses withClue to add file:line context to assertions.
        // In Rust, assert_eq! automatically shows file and line on failure,
        // so I'll just add helpful hex-encoded block hashes for debugging.
        assert_eq!(
            last_finalized_block_hash,
            expected.block_hash,
            "Last finalized block mismatch\nExpected: {}\nGot: {}",
            hex::encode(&expected.block_hash),
            hex::encode(&last_finalized_block_hash)
        );
    }

    // Bonds function: _.map(pk => pk -> 10L).toMap
    fn bonds_function(validators: Vec<PublicKey>) -> HashMap<PublicKey, i64> {
        validators.into_iter().map(|pk| (pk, 10i64)).collect()
    }

    let parameters = GenesisBuilder::build_genesis_parameters_with_defaults(
        Some(bonds_function),
        None, // Use default validatorsNum = 4
    );

    let genesis = GenesisBuilder::new()
        .build_genesis_with_parameters(Some(parameters))
        .await
        .expect("Failed to build genesis");

    let mut nodes = TestNode::create_network(genesis.clone(), 3, None, None, None, None)
        .await
        .unwrap();

    let deploy_datas: Vec<_> = (0..=7)
        .map(|i| {
            construct_deploy::basic_deploy_data(
                i,
                None,
                Some(genesis.genesis_block.shard_id.clone()),
            )
            .unwrap()
        })
        .collect();

    let block1 = TestNode::propagate_block_at_index(&mut nodes, 0, &[deploy_datas[0].clone()])
        .await
        .unwrap();

    let block2 = TestNode::propagate_block_at_index(&mut nodes, 1, &[deploy_datas[1].clone()])
        .await
        .unwrap();

    let block3 = TestNode::propagate_block_at_index(&mut nodes, 2, &[deploy_datas[2].clone()])
        .await
        .unwrap();

    let block4 = TestNode::propagate_block_at_index(&mut nodes, 0, &[deploy_datas[3].clone()])
        .await
        .unwrap();

    let _block5 = TestNode::propagate_block_at_index(&mut nodes, 1, &[deploy_datas[4].clone()])
        .await
        .unwrap();

    assert_finalized_block(&nodes[0], &block1);

    let _block6 = TestNode::propagate_block_at_index(&mut nodes, 2, &[deploy_datas[5].clone()])
        .await
        .unwrap();

    assert_finalized_block(&nodes[0], &block2);

    let _block7 = TestNode::propagate_block_at_index(&mut nodes, 0, &[deploy_datas[6].clone()])
        .await
        .unwrap();

    assert_finalized_block(&nodes[0], &block3);

    let _block8 = TestNode::propagate_block_at_index(&mut nodes, 1, &[deploy_datas[7].clone()])
        .await
        .unwrap();

    assert_finalized_block(&nodes[0], &block4);
}

/// This test verifies that finalization advances monotonically (block number never
/// decreases) during round-robin block production with multi-parent merging, and
/// that all validators agree on the last finalized block at the end.
#[tokio::test]
async fn multi_parent_casper_should_advance_finalization_monotonically_in_round_robin() {
    fn bonds_function(validators: Vec<PublicKey>) -> HashMap<PublicKey, i64> {
        validators.into_iter().map(|pk| (pk, 10i64)).collect()
    }

    let parameters = GenesisBuilder::build_genesis_parameters_with_defaults(
        Some(bonds_function),
        None, // Use default validatorsNum = 4
    );

    let genesis = GenesisBuilder::new()
        .build_genesis_with_parameters(Some(parameters))
        .await
        .expect("Failed to build genesis");

    let mut nodes = TestNode::create_network(genesis.clone(), 3, None, None, None, None)
        .await
        .unwrap();

    let shard_id = genesis.genesis_block.shard_id.clone();

    let deploy_datas: Vec<_> = (0..=20)
        .map(|i| construct_deploy::basic_deploy_data(i, None, Some(shard_id.clone())).unwrap())
        .collect();

    // Round-robin block production across 3 validators with propagation.
    // With multi-parent merging, each block may have multiple parents
    // (one from each validator's latest message).
    let _block1 = TestNode::propagate_block_at_index(&mut nodes, 0, &[deploy_datas[0].clone()])
        .await
        .unwrap();

    let _block2 = TestNode::propagate_block_at_index(&mut nodes, 1, &[deploy_datas[1].clone()])
        .await
        .unwrap();

    let _block3 = TestNode::propagate_block_at_index(&mut nodes, 2, &[deploy_datas[2].clone()])
        .await
        .unwrap();

    let _block4 = TestNode::propagate_block_at_index(&mut nodes, 0, &[deploy_datas[3].clone()])
        .await
        .unwrap();

    let _block5 = TestNode::propagate_block_at_index(&mut nodes, 1, &[deploy_datas[4].clone()])
        .await
        .unwrap();

    // After 5 blocks in round-robin with 3 validators (equal bonds),
    // finalization should have advanced past genesis.
    let lfb_after5 = nodes[0].casper.last_finalized_block().await.unwrap();

    let _block6 = TestNode::propagate_block_at_index(&mut nodes, 2, &[deploy_datas[5].clone()])
        .await
        .unwrap();

    let lfb_after6 = nodes[0].casper.last_finalized_block().await.unwrap();

    let _block7 = TestNode::propagate_block_at_index(&mut nodes, 0, &[deploy_datas[6].clone()])
        .await
        .unwrap();

    let lfb_after7 = nodes[0].casper.last_finalized_block().await.unwrap();

    let _block8 = TestNode::propagate_block_at_index(&mut nodes, 1, &[deploy_datas[7].clone()])
        .await
        .unwrap();

    let lfb_after8 = nodes[0].casper.last_finalized_block().await.unwrap();

    // Verify finalization advances monotonically (block number never decreases)
    assert!(
        lfb_after6.body.state.block_number >= lfb_after5.body.state.block_number,
        "LFB block number should not decrease: after6={} < after5={}",
        lfb_after6.body.state.block_number,
        lfb_after5.body.state.block_number
    );
    assert!(
        lfb_after7.body.state.block_number >= lfb_after6.body.state.block_number,
        "LFB block number should not decrease: after7={} < after6={}",
        lfb_after7.body.state.block_number,
        lfb_after6.body.state.block_number
    );
    assert!(
        lfb_after8.body.state.block_number >= lfb_after7.body.state.block_number,
        "LFB block number should not decrease: after8={} < after7={}",
        lfb_after8.body.state.block_number,
        lfb_after7.body.state.block_number
    );

    // Finalization progression can be delayed under high merge pressure.
    // Continue producing in round-robin and require eventual advancement
    // while preserving monotonic LFB movement.
    let mut latest_lfb = lfb_after8.clone();
    for step in 0..=12 {
        if latest_lfb.block_hash != genesis.genesis_block.block_hash {
            break;
        }
        let producer_idx = ((step + 2) % 3) as usize;
        let deploy_idx = 8 + step as usize;
        let _ = TestNode::propagate_block_at_index(
            &mut nodes,
            producer_idx,
            &[deploy_datas[deploy_idx].clone()],
        )
        .await
        .unwrap();
        let next_lfb = nodes[0].casper.last_finalized_block().await.unwrap();
        assert!(
            next_lfb.body.state.block_number >= latest_lfb.body.state.block_number,
            "LFB block number should not decrease during extended round-robin: next={} < prev={}",
            next_lfb.body.state.block_number,
            latest_lfb.body.state.block_number
        );
        latest_lfb = next_lfb;
    }

    assert!(
        latest_lfb.block_hash != genesis.genesis_block.block_hash,
        "LFB remained at genesis after extended round-robin production"
    );

    // Verify all validators agree on finalization
    let lfb_node1 = nodes[1].casper.last_finalized_block().await.unwrap();
    let lfb_node2 = nodes[2].casper.last_finalized_block().await.unwrap();
    assert_eq!(
        lfb_node1.block_hash, latest_lfb.block_hash,
        "Node1 LFB should match Node0 LFB"
    );
    assert_eq!(
        lfb_node2.block_hash, latest_lfb.block_hash,
        "Node2 LFB should match Node0 LFB"
    );
}
